helix-core/src/comment.rs

```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
     Transaction::change(doc, changes.into_iter())
 }
 
+fn find_last_non_whitespace_char(text: RopeSlice) -> Option<usize> {
+    (0..text.len_chars()).rev().find(|&i| {
+        text.get_char(i)
+            .map(|ch| !ch.is_whitespace())
+            .unwrap_or(false)
+    })
+}
+
+#[derive(Debug, PartialEq)]
+struct CommentChange {
+    range: Range,
+    open_pos: usize,
+    close_pos: usize,
+    open_margin: usize,
+    close_margin: usize,
+}
+
+fn find_block_comments(
+    open: &str,
+    close: &str,
+    text: RopeSlice,
+    selection: &Selection,
+) -> (bool, Vec<CommentChange>) {
+    let mut commented = true;
+    let mut comment_changes = Vec::with_capacity(selection.len());
+    for range in selection {
+        let selection_slice = range.slice(text);
+        if let (Some(open_pos), Some(close_pos)) = (
+            find_first_non_whitespace_char(selection_slice),
+            find_last_non_whitespace_char(selection_slice),
+        ) {
+            let len = selection_slice.len_chars();
+
+            // selection can be shorter than open_pos + open len
+            let open_fragment = Cow::from(
+                selection_slice.slice(open_pos..std::cmp::min(open_pos + open.len(), len)),
+            );
+            let close_fragment = Cow::from(
+                selection_slice
+                    .slice(close_pos - std::cmp::min(close.len() - 1, close_pos)..close_pos + 1),
+            );
```

Comment:
no need to create a `Cow` you can compare `RopeSlice` to `str` directly. Also instead of bounding the size, you can simply check upfront that slice is large enough.

You also mix up byte and char indices here. Ropey and helix use char indices by default while strings use byte indices. I this case you should probably just calculate the char len of `open` and `close` at the start of the function and use char indices everywhere. This won't make a difference for ASCII comments but we should correctly handle non-ascii tokens even if they are rare in practice.

Reply:
ah yes this pr was pretty much the first thing I wrote in rust so I just blindly copied this from the line commenting and modified it work for block commenting, looks like there should be a follow up pr to improve the line commenting implementation to take this into account as well.

Reply:
That would be nice, also split selection at whitespace could be improved